homeassistant/components/google_assistant/helpers.py

```diff
@@ -323,13 +324,36 @@ def __init__(self, hass):
         """Initialize a configuration store."""
         self._hass = hass
         self._store = Store(hass, self._STORAGE_VERSION, self._STORAGE_KEY)
-        self._data = {STORE_AGENT_USER_IDS: {}}
+        self._data = None
+
+    async def async_initialize(self):
+        """Finish initializing the ConfigStore."""
+        if (data := await self._store.async_load()) is None:
+            data = {
+                STORE_AGENT_USER_IDS: {},
+                STORE_GOOGLE_LOCAL_WEBHOOK_ID: webhook.async_generate_id(),
+            }
+            await self._store.async_save(data)
+
+        self._data = data
+
+        if STORE_GOOGLE_LOCAL_WEBHOOK_ID not in self._data:
+            self._data = {
+                **self._data,
+                STORE_GOOGLE_LOCAL_WEBHOOK_ID: webhook.async_generate_id(),
```

Comment:
This is not correct. The manual configured Google Assistant supports multiple users. So in that case each user needs to have their own webhook ID.

Reply:
Could you please explain how multiple users a re currently supported? I did some digging and found that the user agents are used in two cases:
1. to pass on to service calls that perform the actual actions for each trait.
2. to 'support' local fulfillment.

I am uncertain whether the first use case is actually implemented, but the use case is still supported as the user id gets passed along nicely as is for cloud calls, and in the local fulfillment scenario it is always the first user agent id, as we do not know which it was.

For the second use case it does not make sense to support multiple user agent ids. This is because in that scenario we are providing the value of user agent id to what should be the home assistant device id in the IDENTIFY request and the proxy devicy id in the SYNC request. ( I do not know how to link to code here but these it can be found in smart_home.py line 267, and smart_home.py line 89 which goes to helpers.py 530 respectively ). To me it makes no sense to use Home Assistant user ID values for this, as there is only one Home Assistant device. I would propose using one unique id here.

I will make these changes and push them to this pull request. Note that this most likely will make the entire storing of the user agent ids obsolete. I will leave it in for now as the cloud integration makes use of it, and I think it is outside the scope of this pull request to refactor the entire component (I am willing to open up a new pull request for that though). If you think it should be included in this request let me know, and I will make the changes here.

Reply:
The docs mention `https://[YOUR HOME ASSISTANT URL:PORT]/auth/authorize` -> when you link the skill in your Google Assistant, you are logging into Home Assistant. Multiple people can do this, connected to different "Homes" in Google. That's why the code mentions plural user agent ids and why there is a method `async_report_state_all`.

And yes, it's unnecessarily complicated but that's how it is now. 

To identify which user agent belongs to which local communication, you can register unique webhooks for each user. 

You can't just use the first user ID, because that means that all actions the 2nd user does via Google local is attributed to the first.

If you want to limit the local SDK to a single user you need to make sure that the local SDK information is not synced back to Google for the other users.

Reply:
So the situation you are describe is as follows (please correct me if I am wrong):

There is one home assistant device.
There is two (or more) google home devices belonging to different users/homes that are both connected to this one home assistant device.

Both these devices have logged into the the /auth/authorize endpoint.

In this scenario both google home devices would need to talk to the same home assistant device and as long as the cloud handler already gets the user id, this would still get passed to the later requests in home assistant (so that would be in http.py line 240 for getting the user id).

But in this scenario it does not make sense to sync the same devices twice (once for each user) as the data we currently send (without local sdk enabled) is exactly the same as it is not dependent on the user_agent_id (see the serialize function in the sync - helpers.py line 475 -  as only line 515 uses the agent_user_id and this is behind an if for local sdk).

Or is the scenario you mean that there are multiple home assistant devices in different homes that somehow use the same google setup and /auth/authorize endpoint?

Reply:
In the case that there is multiple home assistant devices, the current multiple user agent ids also would not make sense to me as those users are local to a single instance as far as I am aware.

Reply:
> To identify which user agent belongs to which local communication, you can register unique webhooks for each user.

In home assistant this is possible yes, but this gets send to Google in the customData field (when local sdk is enabled) and this field will be overwritten by any later sync on that same deviceId, so google at any one point in time has at most 1 webhookId stored for each synced device.

Reply:
> I am uncertain whether the first use case is actually implemented, but the use case is still supported as the user id gets passed along nicely as is for cloud calls, and in the local fulfillment scenario it is always the first user agent id, as we do not know which it was.

I did some more testing and in the case of local sdk where a user id is specified that user id is also written to the Logbook in the frontend, so the changes still support showing which user performed the action if that worked before :).

Reply:
Okay, I think I now understand how it is done, the home assistant system is registered twice with all devices also registering twice, therefore you can have multiple customData for the same device as they live within another userAgentId (they act as if they are different devices). This would mean that indeed the webhook can and needs to be bound to a specific user. So I need to revert plenty of changes and try that approach. The only issue I have there is that the current implentation of AbstractConfig sort of does not really allow for this with its properties.